crates/swc_cli/src/commands/compile.rs

```diff
@@ -1,5 +1,220 @@
+    let base = file_path
+        .parent()
+        .unwrap_or(&default)
+        .as_os_str()
+        .to_string_lossy();
+
+    let dist_absolute_path = out_dir.absolutize()?;
+
+    // These are possible combinations between input to output dir.
+    // cwd: /c/github/swc
+    //
+    // Input
+    // 1. Relative to cwd                   : ./crates/swc/tests/serde/a.js
+    // 2. Relative to cwd, traverse up      : ../repo/some/dir/b.js
+    // 3. Absolute path, relative to cwd: /c/github/swc/crates/swc/tests/serde/a.js
+    // 4. Absolute path, not relative to cwd: /c/github/repo/some/dir/b.js
+    //
+    // OutDir
+    // a. Relative to cwd: ./dist
+    // b. Relative to cwd, traverse up: ../outer_dist
+    // c. Absolute path: /c/github/swc/dist
+    // d. Absolute path, not relative to cwd: /c/github/outer_dist
+    //
+    // It is unclear how to calculate output path when either input or output is not
+    // relative to cwd (2,4 and b,d) and it is UB for now.
+    let base = RelativePath::new(&*base);
+    let output_path = base.to_logical_path(dist_absolute_path).join(
+        // Custom output file extension is not supported yet
+        file_path
+            .with_extension("js")
+            .file_name()
+            .expect("Filename should be available"),
+    );
+
+    Ok(output_path)
+}
+
+fn emit_output(
+    output: &TransformOutput,
+    out_dir: &Option<PathBuf>,
+    file_path: &Path,
+) -> anyhow::Result<()> {
+    if let Some(out_dir) = out_dir {
+        let output_file_path = resolve_output_file_path(out_dir, file_path)?;
+        let output_dir = output_file_path
+            .parent()
+            .expect("Parent should be available");
+
+        if !output_dir.is_dir() {
+            fs::create_dir_all(output_dir)?;
+        }
+
+        fs::write(&output_file_path, &output.code)?;
+
+        if let Some(source_map) = &output.map {
+            let source_map_path = output_file_path.with_extension("js.map");
+            fs::write(source_map_path, source_map)?;
+        }
+    } else {
+        println!(
+            "{}\n{}\n{}",
+            file_path.display(),
+            output.code,
+            output
+                .map
+                .as_ref()
+                .map(|m| m.to_string())
+                .unwrap_or_default()
+        );
+    };
+    Ok(())
+}
+
+impl super::CommandRunner for CompileOptions {
+    fn execute(&self) -> anyhow::Result<()> {
+        let included_extensions = if let Some(extensions) = &self.extensions {
+            extensions.clone()
+        } else {
+            DEFAULT_EXTENSIONS.iter().map(|v| v.to_string()).collect()
+        };
+
+        let files = get_files_list(&self.files, &included_extensions, false)?;
+        let cm = COMPILER.clone();
+
+        if let Some(out_dir) = &self.out_dir {
+            fs::create_dir_all(out_dir)?;
+        }
+
+        files
+            .into_par_iter()
+            .try_for_each_with(cm, |compiler, file_path| {
```

Comment:
I choose naive parallel iterator for processing files, maybe there's better / recommended practices?

Reply:
The parallel iterator is the most ideal one if we have a list of files to process.